File: backend/app/models/portal_sharing.py
# Copyright (c) 2026 Laurent Barbe
# Licensed under the Apache License, Version 2.0

"""Portal collaboration, sharing, and public-link API contracts."""

from datetime import datetime
from typing import Literal, Optional

from pydantic import AwareDatetime, Field

from app.db import PortalAccountRole
from app.models.base import ApiModel
from app.models.portal_storage_spaces import (
    PortalStorageSpaceAccountMemberRole,
    PortalStorageSpaceGrantRole,
    PortalStorageSpaceRole,
)
from app.models.user import UserAvatar


PortalStorageSpaceShareDirection = Literal["with_me", "by_me"]


class PortalStorageSpaceShare(ApiModel):
    id: str
    storage_space_id: str
    storage_space_name: str
    user_id: Optional[int] = None
    email: str
    role: PortalStorageSpaceGrantRole
    direction: PortalStorageSpaceShareDirection
    activity_label: str = "Active"


class PortalStorageSpaceAccessPerson(ApiModel):
    user_id: Optional[int] = None
    email: str
    display_name: Optional[str] = None
    role: PortalStorageSpaceRole
    portal_role: Optional[PortalAccountRole] = None
    access_source: Optional[Literal["owner", "direct", "group", "direct_and_group"]] = None
    avatar: Optional[UserAvatar] = None


class PortalStorageSpaceAccessSummary(ApiModel):
    mode: Literal["private", "all", "restricted"]
    default_account_member_role: Optional[PortalStorageSpaceAccountMemberRole] = None
    owner: Optional[PortalStorageSpaceAccessPerson] = None
    effective_member_count: int = 0
    explicit_shares: list[PortalStorageSpaceShare] = Field(default_factory=list)
    public_link_count: int = 0
    can_manage_access: bool = False
    can_create_public_links: bool = False


class PortalStorageSpaceShareCandidate(ApiModel):
    user_id: int
    email: str
    display_name: Optional[str] = None
    portal_role: PortalAccountRole
    access_source: Literal["direct", "group", "direct_and_group"]
    already_shared: bool = False
    avatar: Optional[UserAvatar] = None


class PortalCollaborator(ApiModel):
    user_id: int
    email: str
    display_name: Optional[str] = None
    portal_role: PortalAccountRole
    access_source: Literal["direct", "group", "direct_and_group"]
    member_since: Optional[datetime] = None
    avatar: Optional[UserAvatar] = None
    can_review_access: bool = False


PortalCollaboratorStorageSpaceAccessSource = Literal[
    "direct",
    "team",
    "owner",
    "project_manager",
]


class PortalCollaboratorStorageSpaceAccess(ApiModel):
    storage_space_id: str
    storage_space_name: str
    role: PortalStorageSpaceRole
    source: PortalCollaboratorStorageSpaceAccessSource
    can_revoke: bool = False


class PortalCollaboratorAccessReview(ApiModel):
    collaborator: PortalCollaborator
    can_request_project_removal: bool = False
    space_accesses: list[PortalCollaboratorStorageSpaceAccess] = Field(default_factory=list)


class PortalCollaboratorTrend(ApiModel):
    window: Literal["month", "week", "day"]
    label: str
    period_start: str
    collaborator_count: int = 0


class PortalCollaboratorSummary(ApiModel):
    collaborator_count: int = 0
    external_access_key_count: int = 0
    trend: Optional[PortalCollaboratorTrend] = None


class PortalCollaboratorsResponse(ApiModel):
    summary: PortalCollaboratorSummary
    collaborators: list[PortalCollaborator] = Field(default_factory=list)


class PortalStorageSpaceSharePayload(ApiModel):
    email: Optional[str] = None
    user_id: Optional[int] = None
    role: PortalStorageSpaceGrantRole


class PortalStorageSpaceShareUpdate(ApiModel):
    role: PortalStorageSpaceGrantRole


class PortalPublicLink(ApiModel):
    id: int
    storage_space_id: str
    storage_space_name: str
    object_key: str
    object_name: str
    url: str
    label: Optional[str] = None
    created_by_email: Optional[str] = None
    created_at: datetime
    expires_at: Optional[datetime] = None
    revoked_at: Optional[datetime] = None
    status: str


class PortalPublicLinkCreate(ApiModel):
    object_key: str = Field(min_length=1)
    label: Optional[str] = Field(default=None, max_length=120)
    expires_at: Optional[AwareDatetime] = None

File: frontend/src/api/portalSharing.ts
/*
 * Copyright (c) 2026 Laurent Barbe
 * Licensed under the Apache License, Version 2.0
 */
import type { PortalAccountRole } from "./accountAccess";
import type { S3AccountSelector } from "./accountParams";
import { withS3AccountParam } from "./accountParams";
import client from "./client";
import type {
  PortalStorageSpaceAccountMemberRole,
  PortalStorageSpaceGrantRole,
  PortalStorageSpaceRole,
} from "./portal";
import type { UserAvatarDescriptor } from "./users";

export type PortalStorageSpaceShareDirection = "with_me" | "by_me";

export type PortalStorageSpaceShare = {
  id: string;
  storage_space_id: string;
  storage_space_name: string;
  user_id?: number | null;
  email: string;
  role: PortalStorageSpaceGrantRole;
  direction: PortalStorageSpaceShareDirection;
  activity_label?: string | null;
};

export type PortalStorageSpaceAccessPerson = {
  user_id?: number | null;
  email: string;
  display_name?: string | null;
  role: PortalStorageSpaceRole;
  portal_role?: PortalAccountRole | null;
  access_source?: "owner" | "direct" | "group" | "direct_and_group" | null;
  avatar?: UserAvatarDescriptor | null;
};

export type PortalStorageSpaceAccessSummary = {
  mode: "private" | "all" | "restricted";
  default_account_member_role?: PortalStorageSpaceAccountMemberRole | null;
  owner?: PortalStorageSpaceAccessPerson | null;
  effective_member_count: number;
  explicit_shares: PortalStorageSpaceShare[];
  public_link_count: number;
  can_manage_access: boolean;
  can_create_public_links: boolean;
};

export type PortalStorageSpaceShareCandidate = {
  user_id: number;
  email: string;
  display_name?: string | null;
  portal_role: PortalAccountRole;
  access_source: "direct" | "group" | "direct_and_group";
  already_shared?: boolean;
  avatar?: UserAvatarDescriptor | null;
};

export type PortalPublicLink = {
  id: number;
  storage_space_id: string;
  storage_space_name: string;
  object_key: string;
  object_name: string;
  url: string;
  label?: string | null;
  created_by_email?: string | null;
  created_at: string;
  expires_at?: string | null;
  revoked_at?: string | null;
  status: string;
};

export async function fetchPortalStorageSpaceAccessSummary(
  accountId: S3AccountSelector,
  spaceId: string,
): Promise<PortalStorageSpaceAccessSummary> {
  const { data } = await client.get<PortalStorageSpaceAccessSummary>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/access-summary`,
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function listPortalShareCandidates(
  accountId: S3AccountSelector,
): Promise<PortalStorageSpaceShareCandidate[]> {
  const { data } = await client.get<PortalStorageSpaceShareCandidate[]>(
    "/portal/share-candidates",
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function listPortalStorageSpaceShareCandidates(
  accountId: S3AccountSelector,
  spaceId: string,
): Promise<PortalStorageSpaceShareCandidate[]> {
  const { data } = await client.get<PortalStorageSpaceShareCandidate[]>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/share-candidates`,
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function grantPortalStorageSpaceShare(
  accountId: S3AccountSelector,
  spaceId: string,
  payload: {
    email?: string;
    user_id?: number;
    role: PortalStorageSpaceGrantRole;
  },
): Promise<PortalStorageSpaceShare> {
  const { data } = await client.post<PortalStorageSpaceShare>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/shares`,
    payload,
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function updatePortalStorageSpaceShare(
  accountId: S3AccountSelector,
  spaceId: string,
  userId: number,
  role: PortalStorageSpaceGrantRole,
): Promise<PortalStorageSpaceShare> {
  const { data } = await client.put<PortalStorageSpaceShare>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/shares/${userId}`,
    { role },
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function revokePortalStorageSpaceShare(
  accountId: S3AccountSelector,
  spaceId: string,
  userId: number,
): Promise<PortalStorageSpaceShare[]> {
  const { data } = await client.delete<PortalStorageSpaceShare[]>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/shares/${userId}`,
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function listPortalStorageSpacePublicLinks(
  accountId: S3AccountSelector,
  spaceId: string,
  options?: { objectKey?: string; includeRevoked?: boolean },
): Promise<PortalPublicLink[]> {
  const baseParams: Record<string, string | boolean> = {};
  if (options?.objectKey) {
    baseParams.object_key = options.objectKey;
  }
  if (options?.includeRevoked) {
    baseParams.include_revoked = true;
  }
  const { data } = await client.get<PortalPublicLink[]>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/public-links`,
    { params: withS3AccountParam(baseParams, accountId) },
  );
  return data;
}

export async function createPortalStorageSpacePublicLink(
  accountId: S3AccountSelector,
  spaceId: string,
  payload: {
    object_key: string;
    label?: string | null;
    expires_at?: string | null;
  },
): Promise<PortalPublicLink> {
  const { data } = await client.post<PortalPublicLink>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/public-links`,
    payload,
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function revokePortalStorageSpacePublicLink(
  accountId: S3AccountSelector,
  spaceId: string,
  linkId: number,
): Promise<PortalPublicLink[]> {
  const { data } = await client.delete<PortalPublicLink[]>(
    `/portal/storage-spaces/${encodeURIComponent(spaceId)}/public-links/${linkId}`,
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}
